File: go/tests/proactive_briefing_test.go
// proactive_briefing_test.go
//
// Verifies ProactiveBriefingService (ported from ProactiveBriefingService.cs):
// TimeUntilNextFire scheduling math, and FireOnce assembly → summarise →
// deliver (with connector fakes). Behavioural — the C# service is IHostedService;
// FireOnce is the public unit the C# also exposes.

package circleai_test

import (
	"context"
	"strings"
	"testing"
	"time"

	circleai "github.com/bhengubv/CircleAI/go"
)

// ── connector fakes ──────────────────────────────────────────────────────────

type fakeCalendar struct {
	id         string
	configured bool
	events     []circleai.BriefingCalendarEvent
}

func (c fakeCalendar) ProviderID() string { return c.id }
func (c fakeCalendar) IsConfigured() bool { return c.configured }
func (c fakeCalendar) ListEvents(context.Context, time.Time, time.Time) ([]circleai.BriefingCalendarEvent, error) {
	return c.events, nil
}

type fakeEmail struct {
	id         string
	configured bool
	unread     []circleai.BriefingEmailMessage
}

func (e fakeEmail) ProviderID() string { return e.id }
func (e fakeEmail) IsConfigured() bool { return e.configured }
func (e fakeEmail) ListUnread(context.Context, int) ([]circleai.BriefingEmailMessage, error) {
	return e.unread, nil
}

type fakeNews struct {
	id         string
	configured bool
	items      []circleai.BriefingNewsItem
}

func (n fakeNews) SourceID() string   { return n.id }
func (n fakeNews) IsConfigured() bool { return n.configured }
func (n fakeNews) FetchLatest(context.Context, int) ([]circleai.BriefingNewsItem, error) {
	return n.items, nil
}

type fakeWeather struct {
	id     string
	sample circleai.BriefingWeatherSample
}

func (w fakeWeather) ProviderID() string { return w.id }
func (w fakeWeather) Current(context.Context, float64, float64) (circleai.BriefingWeatherSample, error) {
	return w.sample, nil
}

type captureNotifier struct {
	headline string
	body     string
	address  *string
	called   int
}

func (c *captureNotifier) Deliver(_ context.Context, headline, body string, address *string) error {
	c.headline = headline
	c.body = body
	c.address = address
	c.called++
	return nil
}

// ── TimeUntilNextFire ────────────────────────────────────────────────────────

func TestProactiveBriefing_TimeUntilNextFire(t *testing.T) {
	lat, lon := 0.0, 0.0
	svc := circleai.NewProactiveBriefingService(circleai.ProactiveBriefingOptions{
		FireTimesUTC: []time.Duration{6*time.Hour + 30*time.Minute, 18 * time.Hour},
		Latitude:     &lat,
		Longitude:    &lon,
	}, circleai.ProactiveBriefingDeps{})

	// 05:00 → next fire 06:30 (1h30m away).
	now := time.Date(2026, 7, 8, 5, 0, 0, 0, time.UTC)
	if got := svc.TimeUntilNextFire(now); got != 90*time.Minute {
		t.Errorf("05:00 → next: got %v want 1h30m", got)
	}
	// 07:00 → next fire 18:00 (11h away).
	now = time.Date(2026, 7, 8, 7, 0, 0, 0, time.UTC)
	if got := svc.TimeUntilNextFire(now); got != 11*time.Hour {
		t.Errorf("07:00 → next: got %v want 11h", got)
	}
	// 18:30 → both today are past; next is 06:30 tomorrow (12h away).
	now = time.Date(2026, 7, 8, 18, 30, 0, 0, time.UTC)
	if got := svc.TimeUntilNextFire(now); got != 12*time.Hour {
		t.Errorf("18:30 → next: got %v want 12h", got)
	}

	// A moment within 30s before a fire rolls to the next day (no double-fire).
	now = time.Date(2026, 7, 8, 6, 29, 45, 0, time.UTC) // 15s before 06:30
	got := svc.TimeUntilNextFire(now)
	if got < 11*time.Hour {
		t.Errorf("within 30s guard should skip to 18:00, got %v", got)
	}
}

// ── FireOnce ─────────────────────────────────────────────────────────────────

func TestProactiveBriefing_FireOnce_AssemblesAndDelivers(t *testing.T) {
	ctx := context.Background()
	addr := "+27820000000"
	lat, lon := -29.85, 31.02
	notifier := &captureNotifier{}

	svc := circleai.NewProactiveBriefingService(circleai.ProactiveBriefingOptions{
		Headline:        "Morning",
		DeliveryAddress: &addr,
		Latitude:        &lat,
		Longitude:       &lon,
	}, circleai.ProactiveBriefingDeps{
		Calendars: []circleai.BriefingCalendarConnector{
			fakeCalendar{id: "gcal", configured: true, events: []circleai.BriefingCalendarEvent{
				{Title: "Standup", StartUTC: time.Date(2026, 7, 8, 9, 0, 0, 0, time.UTC)},
			}},
			fakeCalendar{id: "unconfigured", configured: false},
		},
		Emails: []circleai.BriefingEmailConnector{
			fakeEmail{id: "gmail", configured: true, unread: []circleai.BriefingEmailMessage{
				{From: "boss@x.com", Subject: "Roadmap"},
			}},
		},
		News: []circleai.BriefingNewsSource{
			fakeNews{id: "hn", configured: true, items: []circleai.BriefingNewsItem{{Title: "Go 2 announced"}}},
		},
		Weather: fakeWeather{id: "owm", sample: circleai.BriefingWeatherSample{
			TempC: 21, FeelsLikeC: 20, WindKph: 12, Condition: "Clear",
		}},
		Notifiers: []circleai.IBriefingNotifier{notifier},
		// Summariser echoes a marker + the raw context so we can assert assembly.
		Summariser: func(_ context.Context, prompt string) (string, error) {
			return "SUMMARY::" + prompt, nil
		},
	})

	if err := svc.FireOnce(ctx); err != nil {
		t.Fatalf("FireOnce: %v", err)
	}
	if notifier.called != 1 {
		t.Fatalf("notifier called %d times, want 1", notifier.called)
	}
	if notifier.headline != "Morning" {
		t.Errorf("headline: got %q", notifier.headline)
	}
	if notifier.address == nil || *notifier.address != addr {
		t.Errorf("address: got %v", notifier.address)
	}
	body := notifier.body
	for _, want := range []string{"SUMMARY::", "### Calendar (gcal)", "Standup", "### Unread email (gmail)", "boss@x.com", "### News (hn)", "Go 2 announced", "### Weather (owm)", "Clear"} {
		if !strings.Contains(body, want) {
			t.Errorf("briefing body missing %q:\n%s", want, body)
		}
	}
	// The unconfigured calendar must not appear.
	if strings.Contains(body, "unconfigured") {
		t.Errorf("unconfigured connector leaked into body:\n%s", body)
	}
}

func TestProactiveBriefing_FireOnce_NoSignalsNoDelivery(t *testing.T) {
	ctx := context.Background()
	notifier := &captureNotifier{}
	svc := circleai.NewProactiveBriefingService(circleai.ProactiveBriefingOptions{},
		circleai.ProactiveBriefingDeps{
			// All connectors unconfigured / empty.
			Calendars: []circleai.BriefingCalendarConnector{fakeCalendar{id: "c", configured: false}},
			Notifiers: []circleai.IBriefingNotifier{notifier},
		})
	if err := svc.FireOnce(ctx); err != nil {
		t.Fatalf("FireOnce: %v", err)
	}
	if notifier.called != 0 {
		t.Errorf("no signals should mean no delivery, got %d calls", notifier.called)
	}
}

func TestProactiveBriefing_NoSummariserDeliversRawContext(t *testing.T) {
	ctx := context.Background()
	notifier := &captureNotifier{}
	svc := circleai.NewProactiveBriefingService(circleai.ProactiveBriefingOptions{},
		circleai.ProactiveBriefingDeps{
			News: []circleai.BriefingNewsSource{
				fakeNews{id: "rss", configured: true, items: []circleai.BriefingNewsItem{{Title: "Headline A"}}},
			},
			Notifiers: []circleai.IBriefingNotifier{notifier},
			// No summariser → raw context delivered.
		})
	if err := svc.FireOnce(ctx); err != nil {
		t.Fatalf("FireOnce: %v", err)
	}
	if notifier.called != 1 {
		t.Fatalf("expected 1 delivery, got %d", notifier.called)
	}
	if !strings.Contains(notifier.body, "### News (rss)") || !strings.Contains(notifier.body, "Headline A") {
		t.Errorf("raw-context body: %q", notifier.body)
	}
	if strings.HasPrefix(notifier.body, "SUMMARY::") {
		t.Errorf("no summariser should mean no summary prefix: %q", notifier.body)
	}
}

func TestProactiveBriefing_StartStop(t *testing.T) {
	ctx := context.Background()
	svc := circleai.NewProactiveBriefingService(circleai.ProactiveBriefingOptions{
		// A far-future fire time so the loop just sleeps.
		FireTimesUTC: []time.Duration{23 * time.Hour},
	}, circleai.ProactiveBriefingDeps{})
	if err := svc.Start(ctx); err != nil {
		t.Fatalf("Start: %v", err)
	}
	// Start is idempotent.
	if err := svc.Start(ctx); err != nil {
		t.Fatalf("second Start: %v", err)
	}
	if err := svc.Stop(ctx); err != nil {
		t.Fatalf("Stop: %v", err)
	}
	if err := svc.Stop(ctx); err != nil {
		t.Fatalf("second Stop: %v", err)
	}
}

File: typescript/tests/proactive_briefing.test.ts
// proactive_briefing.test.ts
//
// Verifies ProactiveBriefingService (ProactiveBriefingService.cs): context
// assembly from calendar / email / news / weather connectors, the LLM summary
// (or raw context fallback), notifier delivery, the "no signals → no fire" path,
// and the next-fire-time computation.

import { describe, it } from 'node:test';
import assert from 'node:assert/strict';
import {
  ProactiveBriefingService,
  DEFAULT_FIRE_TIMES_UTC_MINUTES,
  type ICalendarConnector,
  type IEmailConnector,
  type INewsSource,
  type IWeatherProvider,
  type IBriefingNotifier,
  type IBriefingAIService,
  type BriefingCalendarEvent,
} from '../src/companion/index';

class CapturingNotifier implements IBriefingNotifier {
  deliveries: { headline: string; body: string; address: string | null }[] = [];
  async deliverAsync(headline: string, body: string, address: string | null): Promise<void> {
    this.deliveries.push({ headline, body, address });
  }
}

function calendar(events: readonly BriefingCalendarEvent[], configured = true): ICalendarConnector {
  return {
    providerId: 'google',
    isConfigured: configured,
    async listEventsAsync() {
      return events;
    },
  };
}

function email(configured = true): IEmailConnector {
  return {
    providerId: 'gmail',
    isConfigured: configured,
    async listUnreadAsync() {
      return [{ from: 'Boss', subject: 'Q3 numbers' }];
    },
  };
}

function news(): INewsSource {
  return {
    sourceId: 'hn',
    isConfigured: true,
    async fetchLatestAsync() {
      return [{ title: 'Rust 2.0 released' }];
    },
  };
}

function weather(): IWeatherProvider {
  return {
    providerId: 'openmeteo',
    async currentAsync() {
      return { tempC: 21.6, feelsLikeC: 20.4, windKph: 12.9, condition: 'Cloudy' };
    },
  };
}

describe('ProactiveBriefingService.fireOnceAsync — context assembly', () => {
  it('assembles calendar + email + news + weather and delivers via notifier', async () => {
    const notifier = new CapturingNotifier();
    const startUtc = new Date(Date.UTC(2026, 6, 8, 9, 0, 0));
    const svc = new ProactiveBriefingService(
      { latitude: -26, longitude: 28, headline: 'Morning' },
      {
        calendars: [calendar([{ title: 'Standup', location: 'Zoom', startUtc }])],
        emails: [email()],
        news: [news()],
        weather: weather(),
        notifiers: [notifier],
        // No AI → summary is the raw context.
      },
    );
    await svc.fireOnceAsync();
    assert.equal(notifier.deliveries.length, 1);
    const body = notifier.deliveries[0].body;
    assert.ok(body.includes('### Calendar (google)'));
    assert.ok(body.includes('Standup @ Zoom'));
    assert.ok(body.includes('### Unread email (gmail)'));
    assert.ok(body.includes('- Boss: Q3 numbers'));
    assert.ok(body.includes('### News (hn)'));
    assert.ok(body.includes('- Rust 2.0 released'));
    assert.ok(body.includes('### Weather (openmeteo)'));
    // F0 rounding: 21.6→22, 20.4→20, 12.9→13.
    assert.ok(body.includes('- 22°C Cloudy, feels 20°C, wind 13 km/h'));
    assert.equal(notifier.deliveries[0].headline, 'Morning');
  });

  it('routes context through the AI service when one is present', async () => {
    const notifier = new CapturingNotifier();
    const ai: IBriefingAIService = {
      async chatAsync(messages) {
        // The prompt is the last user message; return a canned summary.
        assert.equal(messages[0].role, 'user');
        assert.ok(messages[0].content.includes('### News (hn)'));
        return 'Your day: one unread email, one headline.';
      },
    };
    const svc = new ProactiveBriefingService(
      {},
      { news: [news()], emails: [email()], notifiers: [notifier], ai },
    );
    await svc.fireOnceAsync();
    assert.equal(notifier.deliveries[0].body, 'Your day: one unread email, one headline.');
  });

  it('falls back to raw context if the AI call throws', async () => {
    const notifier = new CapturingNotifier();
    const ai: IBriefingAIService = {
      async chatAsync() {
        throw new Error('LLM offline');
      },
    };
    const svc = new ProactiveBriefingService({}, { news: [news()], notifiers: [notifier], ai });
    await svc.fireOnceAsync();
    assert.ok(notifier.deliveries[0].body.includes('### News (hn)'));
  });

  it('no signals → no fire (notifier not called)', async () => {
    const notifier = new CapturingNotifier();
    // Unconfigured connectors contribute nothing.
    const svc = new ProactiveBriefingService(
      {},
      { calendars: [calendar([], false)], emails: [email(false)], notifiers: [notifier] },
    );
    await svc.fireOnceAsync();
    assert.equal(notifier.deliveries.length, 0);
  });

  it('caps calendar events at 8 and orders them by start time', async () => {
    const notifier = new CapturingNotifier();
    const base = Date.UTC(2026, 6, 8, 6, 0, 0);
    // 10 events out of order; only the earliest 8 should appear.
    const events: BriefingCalendarEvent[] = [];
    for (let i = 9; i >= 0; i--) {
      events.push({ title: `E${i}`, location: null, startUtc: new Date(base + i * 60_000) });
    }
    const svc = new ProactiveBriefingService({}, { calendars: [calendar(events)], notifiers: [notifier] });
    await svc.fireOnceAsync();
    const lines = notifier.deliveries[0].body.split('\n').filter((l) => l.startsWith('- '));
    assert.equal(lines.length, 8);
    // First listed is E0 (earliest); E8/E9 dropped.
    assert.ok(lines[0].includes('E0'));
    assert.ok(!notifier.deliveries[0].body.includes('E9'));
  });

  it('rejects a null options object', () => {
    // @ts-expect-error deliberate null
    assert.throws(() => new ProactiveBriefingService(null), /opts required/);
  });
});

describe('ProactiveBriefingService.timeUntilNextFireMs', () => {
  it('uses the default 06:30 / 18:00 UTC fire times', () => {
    const svc = new ProactiveBriefingService({}, {});
    // At 05:00 UTC the next fire is 06:30 → 90 minutes.
    const now = new Date(Date.UTC(2026, 6, 8, 5, 0, 0));
    const ms = svc.timeUntilNextFireMs(now);
    assert.equal(ms, 90 * 60_000);
    assert.deepEqual(DEFAULT_FIRE_TIMES_UTC_MINUTES, [6 * 60 + 30, 18 * 60]);
  });

  it('rolls to the next day when past the last fire (and > 30s guard)', () => {
    const svc = new ProactiveBriefingService({ fireTimesUtcMinutes: [6 * 60 + 30] }, {});
    // At 06:30:10 (just past 06:30) → rolls to tomorrow's 06:30.
    const now = new Date(Date.UTC(2026, 6, 8, 6, 30, 10));
    const ms = svc.timeUntilNextFireMs(now);
    // ~ 24h minus 10s.
    assert.ok(ms > 23 * 3_600_000);
  });

  it('no fire times → 1 hour', () => {
    const svc = new ProactiveBriefingService({ fireTimesUtcMinutes: [] }, {});
    assert.equal(svc.timeUntilNextFireMs(new Date()), 3_600_000);
  });
});
